### PR TITLE
Harden superblock cache deserialization

### PR DESCRIPTION
**Pull request by derobins**

### CODE REVIEW

**Review by byrnHDF — APPROVED** [review]

**Review by lrknox — APPROVED** [review]

File: src/H5Fsuper_cache.c [hunk 1]
```diff
@@ -135,53 +133,59 @@ H5FL_EXTERN(H5F_super_t);
...
-    HDassert(((size_t)(image - (const uint8_t *)*image_ref)) == H5F_SUPERBLOCK_FIXED_SIZE);
+    /* Size check */
+    if (((size_t)(image - (const uint8_t *)*image_ref)) != H5F_SUPERBLOCK_FIXED_SIZE)
+        HGOTO_ERROR(H5E_FILE, H5E_BADVALUE, FAIL, "bad superblock (fixed) size")
 
     /* Determine the size of addresses & size of offsets, for computing the
      * variable-sized portion of the superblock.
      */
     if (sblock->super_vers < HDF5_SUPERBLOCK_VERSION_2) {
+        if (H5_IS_BUFFER_OVERFLOW(image, 6, end))
```
Comment (jhendersonHDF) [inline src/H5Fsuper_cache.c:177]: 2 instead of 6?

File: src/H5Fsuper_cache.c [hunk 2]
```diff
@@ -430,134 +421,122 @@ H5F__cache_superblock_deserialize(const void *_image, size_t H5_ATTR_NDEBUG_UNUS
...
 
-        /* Check whether the image pointer is out of bounds */
-        if (H5_IS_BUFFER_OVERFLOW(image, sizeof(uint32_t), image_end))
-            HGOTO_ERROR(H5E_FILE, H5E_OVERFLOW, NULL, "image pointer is out of bounds")
-
         /* File status flags (not really used yet) */
+        if (H5_IS_BUFFER_OVERFLOW(image, 4, end))
+            HGOTO_ERROR(H5E_FILE, H5E_OVERFLOW, NULL, "image pointer is out of bounds")
         UINT32DECODE(image, status_flags);
         HDassert(status_flags <= 255);
```
Comment (jhendersonHDF) [inline src/H5Fsuper_cache.c]: Convert to regular check instead of assertion

File: src/H5Fsuper_cache.c (hunk 1 above)
Comment (derobins) [reply src/H5Fsuper_cache.c:177]: If you look at the layout for superblock versions 0 and 1, you'll see that there's a set of 7 bytes after the superblock version number. The following code reads the 5th and 6th bytes.

File: src/H5Fsuper_cache.c (hunk 2 above)
Comment (derobins) [reply src/H5Fsuper_cache.c]: Oops!

**Pull request merged**